### PR TITLE
Added AnyLeaf H7 FC target

### PR DESCRIPTION
**Pull request by David-OConnor**
LMK if anyone would like a sample.

Note: Currently the status LED is not working; LED on PC13, setup with this; the rest works.
```
# LED
PC13 LED1 OUTPUT LOW GPIO(91) # amber
define HAL_GPIO_A_LED_PIN 91
```

### CODE REVIEW

**Comment by Hwurzburg** [comment]
needs readme..see KakuteH7Wing for example format..do you have a link to photo/bd layout? the board ID is already assigned to PodMAN H7 is that you?if not get a separate PR in asap to reserve a new id in the board_types.txt file for this board
always good to send Tridge a sample and Andy Piper and sometimes me (especially if Plane is a target for the board)

**Comment by David-OConnor** [comment]
PR in; ID changed here; pic attached. I'll reach out on Discord for mailing addresses. Stby for ReadMe.

h7

**Comment by David-OConnor** [comment]
Readme added.

File: libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md
Comment (Hwurzburg) [inline libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md:58]: needs grouping info and which outputs are bidir dshot capable

File: libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/defaults.parm
Comment (Hwurzburg) [inline libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/defaults.parm:25]: this should be the default unless you have a special DroneCAN GPS that needs config from the autopilot instead of locally

File: libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/defaults.parm
Comment (Hwurzburg) [inline libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/defaults.parm:19]: this will cause issues with Walksnail and DJI WTFOS....should leave this at default I think

File: libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md
Comment (Hwurzburg) [inline libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md:35]: needs dma capability addeds for each serial port

File: libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md
Comment (Hwurzburg) [inline libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md:36]: this is incorrect according to your def

File: libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md
Comment (Hwurzburg) [inline libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md:37]: this should be DJI connector according to your defaults

File: libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md
Comment (Hwurzburg) [inline libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md:38]: this is not setup as ESC telem by default, maybe should be?

File: libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/defaults.parm
Comment (Hwurzburg) [inline libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/defaults.parm:4]: I would not advise this.....switches are very personal

File: libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/defaults.parm
Comment (Hwurzburg) [inline libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/defaults.parm:5]: if this is ELRS you may want the RC_OPTIONS bit 13

File: libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/defaults.parm
Comment (Hwurzburg) [inline libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/defaults.parm:7]:
```suggestion
RSSI_TYPE 3 # RC provided RSSI
```

File: libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md
Comment (Hwurzburg) [inline libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md:64]: AP does not support serial compasses only I2C and DroneCAN

File: libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md
Comment (Hwurzburg) [inline libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md:28]: should include images and link in this readme....and all pins are not labeled on the PCB..where is TX2 and RX2 (DJI connector)...same for ESC connector telem pin (RX3)..

File: libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md
Comment (Hwurzburg) [inline libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md:25]: weight and dimensions?

**Review by Hwurzburg — CHANGES_REQUESTED** [review]

File: libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md
Comment (David-OConnor) [reply libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md:58]: What is that? Thanks!

File: libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/defaults.parm
Comment (David-OConnor) [reply libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/defaults.parm:19]: I can confirm that this is the correct setting for WTFOS, and O3.

File: libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md
Comment (David-OConnor) [reply libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md:35]: How do I do that? Isn't that handled by AP? (They are all obviously DMA capable)

File: libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/defaults.parm
Comment (David-OConnor) [reply libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/defaults.parm:5]: Thanks; uncommented the commented out RC_OPTIONS 512 line; that sets the suppress ELRS warnings bit.

File: libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md
Comment (David-OConnor) [reply libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md:40]: DJI is the OSD. What change do you propose?

File: libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md
Comment (David-OConnor) [reply libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md:41]: Good call!

File: libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md
Comment (Hwurzburg) [reply libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md:61]: look at KakuteH7Wing readme

File: libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md
Comment (Hwurzburg) [reply libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md:38]: first you have to include the bootloader files in this PR....it wont build without them....then you  run a waf configure and look at the hw.dat file produced to see what DMA channels get assigned ...also needed to assure that bidir requests are implemented for outputs in the configure....shared channels in bi-dir is a no-no

File: libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/defaults.parm
Comment (David-OConnor) [reply libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/defaults.parm:14]: Also, I recognize that this is a subjective thing; for my own use case and the intended use case, it makes sense to set this default. For the general use case, it is better to leave these to the AP defaults. I think, based on the the fact that this FC has no analog OSD chip, and a DJI-format external port, it makes sense to have it set up by default for MSP Displayport OSD.

A relevant question, as you pointed out, is "Do HDZero and WalkSnail also use MSP DisplayPort (Like WTF OS and O3). I leave this to you or another AP maintainer to make the call; my vote is keep it set up for MSP Displayport like this.

File: libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/defaults.parm
Comment (Hwurzburg) [reply libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/defaults.parm:5]: not the same thing

File: libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md
Comment (David-OConnor) [reply libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md]: I added some clarification text; LMK how you'd like to proceed. I can draft up some images if you'd like, but I think now after the updates, they are probably not required in conjunction with the readme, datasheet, and silkscreen text. Your call.

File: libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md
Comment (David-OConnor) [reply libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md:61]: Of note, I added to the readme details about how the motor pins are designed to be mapped to timers.

Comment (David-OConnor) [reply libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md:61]: It looks like the Kakute's readme uses the motor definition almost C+P from hwdef.dat. Given that it includes information that is not really relevant relevant to the hardware (`PWM(x)` and `GPIO(y)` are AP constructs, and are not directly related to the PCB, and the `TIMx_CH7 TIMx` is also not really readme material and repeats the tim numbers.

File: libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/defaults.parm
Comment (Hwurzburg) [reply libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/defaults.parm:5]: bit 13 is needed....and bit 9 as you had....but bit 9 is subjective....bit 13 is required for ELRS

File: libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md
Comment (David-OConnor) [reply libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md:38]: Are you referring to `hwdef-bl.dat`? That is in the PR. (If it's something else, LMK!)

I'm still not quite sure how to follow those instructions, but I can tell you that these are all on General Purpose STM32 timers; that motors 1-4 are on the same timer, and motors 5/6 and 7/8 are each on another timer. They are all capable of Burst DMA, and bidirectional DSHOT.

File: libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md
Comment (Hwurzburg) [reply libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md:71]: line not changed

File: libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/defaults.parm
Comment (David-OConnor) [reply libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/defaults.parm:4]: Normally, I'd agree, but ELRS hard-sets arm to channel 5, and not using airmode is IMO dangerous in any flight mode.

**Comment by Hwurzburg** [comment]
some review comments still need addressing....need bootloaders and final commit squash and subsystems split of commits

File: libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md
Comment (David-OConnor) [reply libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md]: Oop! Done.

File: libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md
Comment (David-OConnor) [reply libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md:61]: Sounds good; will make the change. How do I know which to tag as BIDIR?

Comment (Hwurzburg) [reply libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md:61]: just get bootloader into PR, then we can do a test build and see how DMA is being allocated...it MAY be possible to have all outputs BIDIR ....maybe not...a shared DMA channel is just that...only one hdwr unit can make use of it at a time

File: libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md
Comment (Hwurzburg) [reply libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md]: when I get a chance in the next few days...I will pull this down when you think you've done everything you can and make some changes and push back up

File: libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md
Comment (David-OConnor) [reply libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md:38]: Where does this place the `hw.dat` Unable to find after running `./waf configure AnyleafH7` and using `grep`.

Comment (Hwurzburg) [reply libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md:38]: Screenshot from 2024-01-18 10-28-59

Comment (David-OConnor) [reply libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md:38]: nvm; found it! `/build/board name`

Comment (David-OConnor) [reply libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md:38]: Here's the part relevant to UART and Motors:

```c
SERIAL_ORDER OTG1 USART1 USART2 USART3 UART4 UART7 UART8
PB6 USART1_TX USART1
PB7 USART1_RX USART1
PA2 USART2_TX USART2
PA3 USART2_RX USART2
PD8 USART3_TX USART3
PD9 USART3_RX USART3
PC10 UART4_TX UART4
PC11 UART4_RX UART4
PB4 UART7_TX UART7
PB3 UART7_RX UART7
PE1 UART8_TX UART8
PE0 UART8_RX UART8
PD0 CAN1_RX CAN1
PD1 CAN1_TX CAN1
PC6  TIM3_CH1 TIM3  PWM(1)  GPIO(50)  BIDIR
PC7  TIM3_CH2 TIM3  PWM(2)  GPIO(51)  BIDIR
PC8  TIM3_CH3  TIM3  PWM(3)  GPIO(52)  BIDIR
PC9  TIM3_CH4  TIM3  PWM(4)  GPIO(53)  BIDIR
PE5  TIM15_CH1  TIM15  PWM(5)  GPIO(54)
PE6  TIM15_CH2  TIM15  PWM(6)  GPIO(55)
PE13 TIM1_CH3 TIM1 PWM(7) GPIO(56)
PE14 TIM1_CH4 TIM1 PWM(8) GPIO(57)
PE9 TIM1_CH1 TIM1 GPIO(58) ALARM
```

File: libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/defaults.parm
Comment (timtuxworth) [reply libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/defaults.parm:13]: HD-Zero supports MSP-Displayport. For HD-Zero OSD_TYPE = 5, SERIALx_PROTOCOL = 42, but MSP_OPTIONS should not be set to 4 for HD-Zero. 
See the wiki https://ardupilot.org/plane/docs/common-msp-osd-overview-4.2.html#displayport-osd

File: libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/defaults.parm
Comment (timtuxworth) [reply libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/defaults.parm:4]: I agree with Henry - there are other ways to do it. I have RC5_OPTION = 0. See my video: https://youtu.be/YO2yA1fmZBs

File: libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/defaults.parm
Comment (David-OConnor) [reply libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/defaults.parm]: Removed MSP_OPTIONS set; great tip.

**Comment by David-OConnor** [comment]
I'm at the point now where I will likely need an assist. The board works with the latest config (other than the LED). Need to do flight tests with the latest updates to config, but other than that, I'm happy on my end. LMK how I can assist re the comments above (motors, serial etc).

**Comment by David-OConnor** [comment]
Diagrams added. LED now working (thanks @IamPete1!)

File: libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/defaults.parm
Comment (Hwurzburg) [reply libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/defaults.parm:4]: and AirMode is potentially dangerous and unexpected, not having it active on arm is not...and many would want a different kind of arm switch...for example ARM/Emergency stop....I will note in the README that I am modifying now and will push shortly, about ELRS arming, but please remove this

**Review by Hwurzburg — CHANGES_REQUESTED** [review]

File: libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/defaults.parm
Comment (Hwurzburg) [inline libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/defaults.parm:22]: remove this liine....you have no default GPS port...remove the SERIAL4 protocol ovverride and it will be that port

**Review by Hwurzburg — CHANGES_REQUESTED** [review]

File: libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/defaults.parm
Comment (Hwurzburg) [inline libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/defaults.parm:21]: remove this...it should be telem if not used for DJI SBUS

**Review by Hwurzburg — CHANGES_REQUESTED** [review]

File: libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/hwdef.dat
Comment (Hwurzburg) [inline libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/hwdef.dat:89]: @andyp1per should review the DMAs for these outputs...some are shared with UARTs....and its possible that outputs 5-8 could also be enabled for BIDIR

**Review by Hwurzburg — CHANGES_REQUESTED** [review]

File: libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/defaults.parm
Comment (David-OConnor) [reply libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/defaults.parm]: Interesting re defaults. Of note, I think CAN is probably a better bet for both GPS and telem. (And you can do telem with the ELRS  onboard as well), but this should be harmless for UART devices.

File: libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md
Comment (Hwurzburg) [reply libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md:42]: sorry its really the hwdef.h file which is only visible after you have bootloaders, which you didnt at this point

File: libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md
Comment (Hwurzburg) [reply libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/README.md]: just pushed readme changes....outputs need to be addressed by @andyp1per and other changes made to defaults

File: libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/hwdef.dat
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/hwdef.dat:89]: You only need to set it once for a 1-2 or 3-4 channel pair. So for example:
```
PC6  TIM3_CH1 TIM3  PWM(1)  GPIO(50)  BIDIR
PC7  TIM3_CH2 TIM3  PWM(2)  GPIO(51)
PC8  TIM3_CH3  TIM3  PWM(3)  GPIO(52)  BIDIR
PC9  TIM3_CH4  TIM3  PWM(4)  GPIO(53)
PE5  TIM15_CH1  TIM15  PWM(5)  GPIO(54) BIDIR
PE6  TIM15_CH2  TIM15  PWM(6)  GPIO(55)
PE13 TIM1_CH3 TIM1 PWM(7) GPIO(56) BIDIR
PE14 TIM1_CH4 TIM1 PWM(8) GPIO(57)
```

Comment (Hwurzburg) [reply libraries/AP_HAL_ChibiOS/hwdef/AnyleafH7/hwdef.dat:106]: need to update in readme which are now bidir also when you make changes like this

**Comment by Hwurzburg** [comment]
@David-OConnor I have made a few changes to avoid DMA sharing (Andy sugggested) and re-instated the alarm buzzer...I have built and loaded on the board you sent me, but you need to confirm dshot operation using your setup....I also squashed, rebased, and put the correct commit messages....once you test and approve, I will tag it for devcall and hopefully merge

**Comment by David-OConnor** [comment]
Sounds great. I flight-tested DSHOT on motors 1-4 in Betaflight yesterday (Valid RPM readings x4), but haven't flight-tested in AP. Will load up the latest and try over the next few days. I appreciate it!

**Comment by David-OConnor** [comment]
Flight tested good today; I'm happy with the latest revision.

**Pull request merged**
